src/main/java/com/aws/greengrass/builtin/services/pubsub/SubscriptionTrie.java

```diff
@@ -69,11 +71,61 @@ public boolean remove(String topic, K cb) {
      * @return if changed after removal
      */
     public boolean remove(String topic, Set<K> cbs) {
-        SubscriptionTrie<K> sub = lookup(topic);
-        if (sub == null) {
+        if (Utils.isEmpty(topic) || Utils.isEmpty(cbs)) {
             return false;
         }
-        return sub.subscriptionCallbacks.removeAll(cbs);
+        SubscriptionTrie<K> current = this;
```

Comment:
nit: I do not disagree with this approach but I think we can accomplish this with one recursive method only. If I understand this correctly, we are passing over the path generated by the topic to remove twice, once in `remove` and again in `pruneRecursively`.

Reply:
Removing(L100) is a necessary step. But `pruneRecursively`(L103) is optional and it is needed only if 'removing' leads to a leaf with no children or no subscriptions.